lib/Dialects/VectorExt/Transform/VectorMaskingUtils.cpp

```diff
@@ -154,3 +154,107 @@ LogicalResult mlir::vector_ext::predicateTiledLoop(OpBuilder &builder,
 
   return success();
 }
+
+/// Materialize masking on a vector.predicate op by generating the masks to
+/// apply to its enclosing operations and erasing the vector.predicate op after
+/// inlining its enclosing operations into its parent region.
+static void maskPredicateOp(OpBuilder &builder, PredicateOp predOp,
+                             SmallVectorImpl<Value> &activeMasks,
+                             const WalkStage &stage) {
+  // Actions before visiting the TruePredicateRegion: Update active mask to be
+  // the predicate mask.
+  if (stage.isBeforeAllRegions()) {
+    activeMasks.push_back(predOp.predicate());
+    return;
+  }
+
+  // Actions after visiting the TruePredicateRegion: Pop the active mask for the
+  // TruePredicateRegion.
+  // if (stage.isAfterRegion(0)) {
+  //   activeMasks.pop_back();
+  //   return;
+  // }
+
+  // TODO: Generate complementary mask for FalsePredicateRegion.
+  // if (stage.isBeforeRegion(1)) {
+  //   return;
+  // }
+
+  // Actions after all the regions: Move operations outside the vector.predicate
+  // regions and remove vector.predicate op.
+  if (stage.isAfterAllRegions()) {
+    // TODO: Move this to isAfterRegion(0);
+    activeMasks.pop_back();
+
+    // Inline truePredicateRegion into parent op (except its terminator).
+    // TODO: Move this to builder.inlineRegionBefore?
+    auto &blocksToMove = predOp.truePredicateRegion().getBlocks();
+    assert(blocksToMove.size() == 1 && "Expected only one block");
+    assert(predOp.getResults().empty() &&
+           "TODO: Support vector.predicate with results");
+    auto &opsToMove = blocksToMove.front().getOperations();
+    moveOperationsBefore(
+        llvm::make_range(opsToMove.begin(), std::prev(opsToMove.end())),
+        predOp);
+    predOp.erase();
+    return;
+  }
+}
+
+/// Traverse `op` and apply masking on all the vector.predicate ops and their
+/// enclosing operations using the strategy `maskGenericOp` to mask generic
+/// operations other than the vector.predicate.
+// TODO: Add an initMask param to pass a mask that is not all-ones?
+LogicalResult mlir::vector_ext::maskVectorPredicateOps(
+    OpBuilder &builder, Operation *op, GenericOpMaskingStrategy maskGenericOp) {
+  OpBuilder::InsertionGuard guard(builder);
+
+  // Initialize stack of active masks. An empty stack means that the active mask
+  // is all-ones (masking is not required).
+  SmallVector<Value, 8> activeMasks;
+  activeMasks.push_back(Value());
+
+  genericWalk(op, [&](Operation *op, const WalkStage &stage) {
```

Comment:
I was pondering whether to invert control to avoid passing WalkStage around (i.e. passins multiple lambdas, e.g. the "doBeforeAllRegions" lambda).

But the way you have it now seems better as we add new ops: we don't want the interleaving of partial control at a distance to become hard to follow; the price is we manipulate WalkStage explicitly.

So more a Q than a comment: have you thought about this (or other tradeoffs) and do you see a reasonable way to avoid leaking the WalkStage ?

One issue in particular is that this is not compatible with pattern-based and will likely be more difficult to integrate in the upcoming "more flexible control" mechanisms that we are iterating on with @ftynse .

Reply:
Note that the `genericWalk` code is a copy-paste from the TF repo (see comment at the top VisitorUtils.cpp/h). I think Rahul Joshi (its owner) had plans to upstream this to core.

IMO, to give this a more pattern-rewriter feeling maybe we could describe the visits for each operation with a class implementing methods like: `doBeforeRegionX`, `doAfterRegionY`, etc. We could then populate a list of these classes with all the visits we want to apply. The main problem I see with this approach is that the visitor algorithm is generic and it should be able to deal with an arbitrary number of regions per op. Mapping each class method to its specific region would require some thought.

I'll ping Rahul about this